packages/headless/src/api/generated-answer/generated-answer-client.ts

```diff
@@ -0,0 +1,166 @@
+export const RETRYABLE_STREAM_ERROR_CODE = 1;
+
+class RetryableError extends Error {}
+class FatalError extends Error {}
+
+interface DispatchActions {
+  updateMessage: (payload: GeneratedAnswerMessagePayload) => void;
+  updateCitations: (payload: GeneratedAnswerCitationsPayload) => void;
+  updateError: (payload: GeneratedAnswerErrorPayload) => void;
+  setIsLoading: (isLoading: boolean) => void;
+  resetAnswer: () => void;
+}
+
+export class GeneratedAnswerAPIClient {
+  private logger: Logger;
+
+  constructor(options: GeneratedAnswerAPIClientOptions) {
+    this.logger = options.logger;
+  }
+
+  streamGeneratedAnswer(
+    params: GeneratedAnswerStreamRequest,
+    actions: DispatchActions
+  ) {
+    const {url, organizationId, streamId, accessToken} = params;
+
+    if (!streamId) {
+      this.logger.error('No stream ID found');
+      return;
+    }
+
+    let retryCount = 0;
+    let timeout: ReturnType<typeof setTimeout> | undefined;
+
+    const retryStream = () => {
+      abortController?.abort();
+      actions.resetAnswer();
+      stream();
+    };
+
+    const refreshTimeout = () => {
+      timeout = resetTimeout(retryStream, timeout, MAX_TIMEOUT);
+    };
+
+    const handleStreamPayload = (
+      payloadType: GeneratedAnswerPayloadType,
+      payload: string
+    ) => {
+      if (payloadType === 'genqa.messageType') {
+        actions.updateMessage(
+          JSON.parse(payload) as GeneratedAnswerMessagePayload
+        );
+      } else if (payloadType === 'genqa.citationsType') {
+        actions.updateCitations(
+          JSON.parse(payload) as GeneratedAnswerCitationsPayload
+        );
+      }
+    };
+
+    const abortController = new AbortController();
+
+    const stream = () =>
+      fetchEventSource(buildStreamingUrl(url, organizationId, streamId).href, {
+        method: 'GET',
+        headers: {
+          Authorization: `Bearer ${accessToken}`,
+        },
+        signal: abortController.signal,
+        async onopen(response) {
+          if (
+            response.ok &&
+            response.headers.get('content-type') === EVENT_STREAM_CONTENT_TYPE
+          ) {
+            return;
+          }
+          const isClientSideError =
+            response.status >= 400 &&
+            response.status < 500 &&
+            response.status !== 429;
+          if (isClientSideError) {
+            throw new FatalError();
+          } else {
+            throw new RetryableError();
+          }
+        },
+        onmessage: (event) => {
+          retryCount = 0;
+          const data: GeneratedAnswerStreamEventData = JSON.parse(event.data);
+          if (data.finishReason === 'ERROR') {
+            clearTimeout(timeout);
+            actions.updateError({
+              message: data.errorMessage,
+              code: data.errorCode,
+            });
+            abortController.abort();
+            return;
+          }
+          if (data.payload && data.payloadType) {
```

Comment:
Would the finishReason be empty or some other value for this case?

Reply:
Would it be more readable then to switch on finishReason ?

Reply:
I personally prefer this as it is flatter and we can `return` and exit the callback.

Reply:
How about using `else if` to make the flow explicit and that the blocks are related ?

Reply:
Sure if you prefer it i can do that for you. Thanks for your review!